.github/workflows/terraform.yml

```
@@ -18,23 +18,40 @@ jobs:
     steps:
       - name: Checkout source code
         uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
-
       - name: Setup Go
         uses: actions/setup-go@0c52d547c9bc32b1aa3301fd7a9cb496313a4491 # v5.0.0
         with:
           go-version-file: 'go.mod'
-
       - name: Setup Terraform
         uses: hashicorp/setup-terraform@b9cd54a3c349d3f38e8881555d616ced269862dd # v3.1.2
         with:
           terraform_wrapper: false
-
       - name: Run unit tests
         run: make test
-
       - name: Run acceptance tests
         env:
           UPCLOUD_USERNAME: ${{ secrets.UPCLOUD_USERNAME }}
           UPCLOUD_PASSWORD: ${{ secrets.UPCLOUD_PASSWORD }}
         if: env.UPCLOUD_PASSWORD != ''
         run: make testacc
+  Cleanup:
+    runs-on: ubuntu-latest
+    needs: Test
+    if: ${{ always() }}
+    steps:
+      - name: Setup Go
+        uses: actions/setup-go@0c52d547c9bc32b1aa3301fd7a9cb496313a4491 # v5.0.0
+      - name: Install upctl
+        run: go install github.com/UpCloudLtd/upcloud-cli/v3/...@latest
+      - name: Purge resources
+        timeout-minutes: 5
+        env:
+          UPCLOUD_USERNAME: ${{ secrets.UPCLOUD_USERNAME }}
+          UPCLOUD_PASSWORD: ${{ secrets.UPCLOUD_PASSWORD }}
+        run: upctl all purge --include *tf-acc-test* --exclude *persistent*
+      - name: List remaining resources
+        env:
+          UPCLOUD_USERNAME: ${{ secrets.UPCLOUD_USERNAME }}
+          UPCLOUD_PASSWORD: ${{ secrets.UPCLOUD_PASSWORD }}
+        run: upctl all list --include *tf-acc-test* --exclude *persistent*
+        if: ${{ always() }}
```

Only needed if purge fails:

```suggestion
        if: ${{ failure() }}
```